keps/0007-dynamic-audit-configuration.md

```diff
@@ -0,0 +1,342 @@
+// AuditConfiguration represents a namespace level audit configuration
+type AuditConfiguration struct {
+    metav1.TypeMeta
+
+    v1.ObjectMeta
+
+    // Policy is the current audit v1beta1 Policy object
+    Policy Policy `json:"policy"`
+
+    // Backends to send events
+    Backends NamespacedBackends `json:"backends"`
+}
+
+// NamespacedBackends holds the configuration for the various namespaced scoped event backends
+type NamespacedBackends struct {
+    // Webhooks holds the webhook backends
+    Webhooks []WebhookBackend `json:"webhooks"`
+}
+```
+
+This object enables multitenant environments to configure audit backends for their specific resources. To start, only webhooks will be accepted as backends in the namespace object as adding file writers could compromise the integrity of the node.   
+
+Example namespaced yaml config:   
+```yaml
+apiVersion: audit.k8s.io/v1beta1
+kind: AuditConfiguration
+metadata:
+  name: <name>
+  namespace: <namespace>
+policy:
+  omitStages:
+    - <stage>
+  rules:
+    - level: <request level>
+      resources:
+      - group: <group>
+        resources: <resources>
+backends:
+  webhooks:
+  - mode: <batch or blocking>
+    initialBackoff: <10s>
+    batch:
+      bufferSize: <400>
+      maxSize: <10000>
+      maxWait: <30s>
+      throttleBurst: <15>
+      throttleEnabled: <true>
+      throttleQPS: <10>
+    config:
+      url: <backend url>
+      service: <optional service name>
+      caBundle: <ca bundle>
+```
+
+### User Stories
+
+#### Story 1
+As a cluster admin I will easily be able to enable the interal auditing features of an existing cluster, and tweak the configurations as necessary. I want to prevent privilege escalation from being able to tamper with a root audit configuration.
+
+#### Story 2
+As a Kubernetes extension developer, I will be able to provide drop in extensions that utilize audit data.
+
+#### Story 3
+As a cluster admin, I will be able configure multiple audit-policies and webhook endpoints to provide independent auditing facilities.
+
+### Implementation Details/Notes/Constraints
+
+Configurations will be cached in memory, as a means of storing the last good configuration. If an update to an existing configuration is unsuccessful, the system will roll back to the previous configuration and log that action, if configured through the api a 400 will be returned. If no previous configuration exists the application of the config will fail, and log the result. Any actions to the audit configuration objects will be hard coded to log at the `level=RequestResponse` to the previous backend and the new backend.
+
+Inherently, apiserver aggregates and HA apiserver setups will work off the same dynamic configuration object. If separate objects are needed they should be configured as static objects on the node and set through the runtime flags. Aggregated servers will implement the same audit handling mechanisms, and a conformance test should be provided as assurance.
+
```

Comment:
I'd like to see a bit more discussion of how this will be implemented, but I'd accept a TODO with a follow up PR. Specifically, the audit policy is currently checked in place in the audit pipeline, but the whole pipeline will need to be refactored to support multiple policies. Or we can just drop multi-policy from this proposal :wink:

Reply:
yeah this just solves the issue of 'I have multiple external audit facilities and want to scope what kind of data is allowed to each'. This is something we can move to a later PR